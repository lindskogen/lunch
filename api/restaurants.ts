import { NowRequest, NowResponse } from "@now/node";
import { fetchAll } from "../lib/lunch-fetcher/fetch-restaurants";

export default async (req: NowRequest, res: NowResponse) => {
  const restaurants = await fetchAll();

  res.setHeader("Cache-Control", "s-maxage=3600, stale-while-revalidate");

  res.json({ restaurants, lastUpdated: new Date().toISOString() });
};
