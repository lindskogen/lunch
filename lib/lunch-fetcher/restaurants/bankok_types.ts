export interface Ingestion {
  price: number;
  type: string;
}

export interface Item {
  id: string;
  name: string;
  preview: string;
  info: string;
  description: string;
  outOfStock: boolean;
  sortOrder: number;
  category: number;
  additionalItems: any[];
  ingestions: Ingestion[];
  itemCategory: string;
}

export interface ItemCategory {
  id: string;
  name: string;
  preview?: any;
  description: string;
  category: number;
}

export interface RestaurantData {
  items: Item[];
  itemCategories: ItemCategory[];
}
