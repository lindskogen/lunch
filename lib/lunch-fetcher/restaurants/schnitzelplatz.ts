import {take, drop} from "lodash-es";
import { Restaurant, FoodItem, RestaurantDayMenu, WeekDay } from "../types";
import { CheerioAPI } from "cheerio";

export const url = "https://schnitzelplatz.se/lunch/";

const firstPartOfSplitOrAll = (str: string, split: string): string => {
  const parts = str.split(split);
  if (parts && parts.length > 1) {
    return parts[0];
  } else {
    return str;
  }
};

export const parseHtml = ($: CheerioAPI): Restaurant => {
  const menus = $(".introbox__container h4")
    .get()
    .map(node => [
      $(node).text().trim(),
      firstPartOfSplitOrAll(
        $(node)
          .next("p")
          .text(),
        "\n"
      )
    ]);

  const numWeekItems = menus.length - 5;

  const weekItems: FoodItem[] = take(menus, numWeekItems).map(
    ([title, name]) => ({
      title: title.toUpperCase(),
      name
    })
  );
  const dayItems = drop(menus, numWeekItems);

  const days: RestaurantDayMenu[] = dayItems.map(([day, name]) => ({
    wday: day as WeekDay,
    items: [{ name }].concat(weekItems)
  }));

  return { name: "Schnitzelplatz Lagerhuset", url, days };
};
