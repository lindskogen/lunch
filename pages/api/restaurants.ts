import { NowRequest, NowResponse } from "@now/node";
import { fetchAll } from "../../lib/lunch-fetcher/fetch-restaurants";

const restaurants = async (req: NowRequest, res: NowResponse) => {
  const restaurants = await fetchAll();

  res.setHeader("Cache-Control", "s-maxage=3600, stale-while-revalidate");

  res.json({ restaurants, lastUpdated: new Date().toISOString() });
};

export default restaurants;
