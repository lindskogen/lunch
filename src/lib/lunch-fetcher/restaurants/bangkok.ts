import { Item, RestaurantData } from "./bangkok_types";
import { groupBy, take } from "lodash-es";
import { filterMap, mapDayNameToWeekDay } from "../lib/utils";
import { RestaurantDayMenu, FoodItem, Restaurant } from "../types";

const QUERY = `
query Query($clientId: String!) {
  items(clientId: $clientId) {
    id
    name
    preview
    info
    description
    outOfStock
    sortOrder
    category
    ingestions {
        price
        type
    }
    itemCategory
  }
  itemCategories(clientId: $clientId) {
      id
      name
      preview
      description
      category
  }
}
`;

const CLIENT_ID = "5c0644cbfd30240bfdb26765";

export const url = "https://www.bkgbg.se/lunch";

export const fetchMeals = (): Promise<RestaurantData> =>
  fetch("https://secure.paidit.se/graphql", {
    method: "POST",
    headers: {
      "content-type": "application/json",
    },
    body: JSON.stringify({
      query: QUERY,
      variables: { clientId: CLIENT_ID },
      operationName: "Query",
    }),
  })
    .then((r) => r.json())
    .then((r) => r.data);

const transformToFoodItem = (item: Item): FoodItem => {
  const isVegetarian = item.description.match(/VEGETARISK/);
  const isVegan = item.description.match(/VEGAN/);

  return {
    name: item.description
      .replace(/\(?VEGETARISK\)? ?/, "")
      .replace(/\(?VEGAN\)? ?/, "")
      .replace(/ \./g, "."),
    title: (
      item.name.replace(/,$/g, "").replace(",", " — ") +
      (isVegetarian ? " — VEGETARISK" : isVegan ? " — VEGANSK" : "")
    ).toUpperCase(),
  };
};

const groupMeals = (data: RestaurantData): RestaurantDayMenu[] => {
  const itemsByItemCategory = groupBy(data.items, (item) => item.itemCategory);

  const days = data.itemCategories.filter((category) =>
    category.name.toLowerCase().includes("lunch")
  );

  return filterMap(days, (day) => {
    const dayName = mapDayNameToWeekDay(day.name);

    if (!dayName) {
      return null;
    }

    return {
      items: take(itemsByItemCategory[day.id], 5).map(transformToFoodItem),
      wday: dayName,
    };
  });
};

export const parseResponse = (restaurantData: RestaurantData): Restaurant => ({
  name: "Bangkok Kitchen",
  url,
  days: groupMeals(restaurantData),
});
