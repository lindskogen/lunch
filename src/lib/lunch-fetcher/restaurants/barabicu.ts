import { isNotNull, filterMap } from "../lib/utils";
import { FoodItem, Restaurant, WeekDay, RestaurantDayMenu } from "../types";

export const url = "https://barabicu.se/";

const weekdayTranslations: WeekDay[] = [
  "Måndag",
  "Tisdag",
  "Onsdag",
  "Torsdag",
  "Fredag",
];
export const parseHtml = ($: CheerioStatic): Restaurant => {
  const menus = $("#todays-lunch li")
    .toArray()
    .map<RestaurantDayMenu>((dayNode, index) => {
      return {
        items: filterMap($("h3", dayNode).toArray(), (elem) => {
          if ($(elem).text() === "" || $(elem.next).text() === "") {
            return null;
          }
          return {
            name: $(elem.next).text(),
            title: $(elem).text().split("•")[0].trim(),
          };
        }),
        wday: weekdayTranslations[index],
      };
    });

  return { name: "Barabicu", url, days: menus };
};
