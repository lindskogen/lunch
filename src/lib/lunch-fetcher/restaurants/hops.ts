import { flatMap, includes, takeWhile, flatten } from "lodash-es";
import { id, isNotNull, isNotZero, isWeekday } from "../lib/utils";
import { FoodItem, Restaurant, WeekDay } from "../types";

export const url = "http://www.hopsbar.se/";

export const parseHtml = ($: CheerioStatic): Restaurant => {
  const menus = $("#Veckansmeny")
    .toArray()
    .map(item => {
      const allNodes = flatMap(item.childNodes, node =>
        $(node)
          .text()
          .trim()
          .split(/\n/)
      )
        .map(s => s.trim())
        .filter(id);

      const weekMealsStartIndex = allNodes.findIndex(
        text => text === "Veckans rätter"
      );

      const weekItems: FoodItem[] = takeWhile(
        allNodes.slice(weekMealsStartIndex + 1),
        text => !includes(text, " kaffe ")
      ).map(food => {
        const foodname = food.trim();
        const matches = foodname.split(':');

        if (matches.length > 1) {
          const [title, name] = matches;
          return { name: name.trim(), title: title.trim().toUpperCase() };
        }

        return { name: foodname };
      });

      const splitIndexes = allNodes
        .map((text, index) => (isWeekday(text) ? index : null))
        .filter(isNotNull);

      let lastIndex = 0;
      const days = splitIndexes
        .filter(isNotZero)
        .map(splitIndex => {
          const slice = allNodes.slice(lastIndex, splitIndex);
          lastIndex = splitIndex;
          return slice;
        })
        .concat([allNodes.slice(lastIndex)]);

      return days
        .filter(([wday]) => isWeekday(wday))
        .map(([wday, ...foods]) => {
          const items = foods.map(food => ({ name: food.trim() }));

          return {
            wday: wday as WeekDay,
            items: takeWhile(
              items,
              ({ name }) => name !== "Veckans rätter"
            ).concat(weekItems)
          };
        });
    });

  return { name: "Hops", url, days: flatten(menus) };
};
