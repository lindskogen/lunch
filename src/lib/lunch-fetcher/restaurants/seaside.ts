import { flatten } from "lodash-es";
import {
  filterMap,
  isNotZero,
  isWeekday,
  mapDayNameToWeekDay,
} from "../lib/utils";
import { Restaurant, WeekDay } from "../types";

export const url = "http://seaside.kvartersmenyn.se/";

export const parseHtml = ($: CheerioStatic): Restaurant => {
  const menus = $(".day .meny")
    .toArray()
    .map((item) => {
      const allNodes = item.childNodes
        .filter((i) => i.tagName === "strong" || i.nodeValue)
        .map((item) => $(item).text());

      const splitIndexes = filterMap(allNodes, (text, index) =>
        isWeekday(text) ? index : null
      );

      let lastIndex = 0;
      const days = splitIndexes
        .filter(isNotZero)
        .map((splitIndex) => {
          const slice = allNodes.slice(lastIndex, splitIndex);
          lastIndex = splitIndex;
          return slice;
        })
        .concat([allNodes.slice(lastIndex)]);

      return filterMap(days, ([wday, ...foods]) => {
        const weekday = mapDayNameToWeekDay(wday);

        if (!weekday) {
          return null;
        }

        const foodsRegex = /^([A-Ã–]+( [0-9])?) ?[\n :]([^0-9:\n]+)/gm;

        const allFoodsForWeekDay = foods.join("\n");

        const parsedFoods = [];

        let res = foodsRegex.exec(allFoodsForWeekDay);
        while (res !== null) {
          const [, title, , name] = res;
          parsedFoods.push({ title, name: name.trim() });
          res = foodsRegex.exec(allFoodsForWeekDay);
        }

        return {
          wday: weekday,
          items: parsedFoods,
        };
      });
    });

  return { name: "Seaside", url, days: flatten(menus) };
};
