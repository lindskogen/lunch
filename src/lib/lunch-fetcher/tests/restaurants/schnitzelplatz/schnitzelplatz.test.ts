import * as cheerio from "cheerio";
import { parseHtml } from "../../../restaurants/schnitzelplatz";

import { mock as mock1 } from "./mock1";

describe("schnitzelplatz", () => {
  describe("when provided with mock 1", () => {
    const result = parseHtml(cheerio.load(mock1));
    it("should work", () => {
      expect(result).toMatchSnapshot();
    });

    it("should have 4 foods per day", () => {
      result.days.forEach((day) => {
        expect(day.items).toHaveLength(4);
      });
    });

    it("should parse 5 days", () => {
      expect(result.days).toHaveLength(5);
    });

    it("should have no empty names", () => {
      result.days.forEach((day) => {
        day.items.forEach((item) => {
          expect(item.name).not.toHaveLength(0);
        });
      });
    });
  });
});
