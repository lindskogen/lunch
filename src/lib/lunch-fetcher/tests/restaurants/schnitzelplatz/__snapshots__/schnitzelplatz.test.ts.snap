// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schnitzelplatz when provided with mock 1 should work 1`] = `
Object {
  "days": Array [
    Object {
      "items": Array [
        Object {
          "name": "Dagens fångst-stroganoff med ris",
        },
        Object {
          "name": "Parmesanpanerad fläskschnitzel med soltorkad tomatkräm och rostad potatis",
          "title": "VECKANS SCHNITZEL",
        },
        Object {
          "name": "Pumpasoppa med smetana",
          "title": "VECKANS SOPPA",
        },
        Object {
          "name": "Bakad spetskål med äppelmajonnäs, haricots verts och rostat matvete",
          "title": "VECKANS VEGETARISKA",
        },
      ],
      "wday": "Måndag",
    },
    Object {
      "items": Array [
        Object {
          "name": "Dagens fångst, fänkålsbakad med vitlökssås och kokt potatis",
        },
        Object {
          "name": "Parmesanpanerad fläskschnitzel med soltorkad tomatkräm och rostad potatis",
          "title": "VECKANS SCHNITZEL",
        },
        Object {
          "name": "Pumpasoppa med smetana",
          "title": "VECKANS SOPPA",
        },
        Object {
          "name": "Bakad spetskål med äppelmajonnäs, haricots verts och rostat matvete",
          "title": "VECKANS VEGETARISKA",
        },
      ],
      "wday": "Tisdag",
    },
    Object {
      "items": Array [
        Object {
          "name": "Dagens fångst, sesambakad med sojamajonnäs och asiatisk kålsallad",
        },
        Object {
          "name": "Parmesanpanerad fläskschnitzel med soltorkad tomatkräm och rostad potatis",
          "title": "VECKANS SCHNITZEL",
        },
        Object {
          "name": "Pumpasoppa med smetana",
          "title": "VECKANS SOPPA",
        },
        Object {
          "name": "Bakad spetskål med äppelmajonnäs, haricots verts och rostat matvete",
          "title": "VECKANS VEGETARISKA",
        },
      ],
      "wday": "Onsdag",
    },
    Object {
      "items": Array [
        Object {
          "name": "Platz höstiga fiskgryta med saffransaioli",
        },
        Object {
          "name": "Parmesanpanerad fläskschnitzel med soltorkad tomatkräm och rostad potatis",
          "title": "VECKANS SCHNITZEL",
        },
        Object {
          "name": "Pumpasoppa med smetana",
          "title": "VECKANS SOPPA",
        },
        Object {
          "name": "Bakad spetskål med äppelmajonnäs, haricots verts och rostat matvete",
          "title": "VECKANS VEGETARISKA",
        },
      ],
      "wday": "Torsdag",
    },
    Object {
      "items": Array [
        Object {
          "name": "Dagens fångst med saffransdoftande vitvinssås, morot och potatisstomp",
        },
        Object {
          "name": "Parmesanpanerad fläskschnitzel med soltorkad tomatkräm och rostad potatis",
          "title": "VECKANS SCHNITZEL",
        },
        Object {
          "name": "Pumpasoppa med smetana",
          "title": "VECKANS SOPPA",
        },
        Object {
          "name": "Bakad spetskål med äppelmajonnäs, haricots verts och rostat matvete",
          "title": "VECKANS VEGETARISKA",
        },
      ],
      "wday": "Fredag",
    },
  ],
  "name": "Schnitzelplatz Lagerhuset",
  "url": "https://schnitzelplatz.se/lunch/",
}
`;
